File: item.py
from stash_preview import ItemDataManager

class Item:
    def __init__(self, name, rarity, position, width, height):
        self.name = name
        self.rarity = rarity
        self.width = width
        self.height = height
        self.position = position

    def __lt__(self, other):
        if self.height != other.height:
            return self.height > other.height
        if self.width != other.width:
            return self.width > other.width
        if self.rarity != other.rarity:
            return self.rarity > other.rarity
        return self.name < other.name
    
    def __eq__(self, other):
        if self and other:
            return self.name == other.name and self.rarity == other.rarity and self.position == other.position
    
    def __hash__(self):
        return hash((self.name, self.rarity, self.position))

    def __repr__(self):
        return f"{self.rarity} {self.name} {self.position} {self.width}X{self.height}"
    
    def to_dict(self):
        return {
            "name": self.name,
            "rarity": self.rarity,
            "position": self.position,
            "width": self.width,
            "height": self.height,
        }

    @staticmethod
    def from_dict(data):
        # "itemId": "DesignDataItem:Id_Item_BloodsapBlade_5001",
        # "slotId": 211,
        item_id = data["itemId"]
        design_str = "DesignDataItem:Id_Item_"
        parts = item_id.replace(design_str, "").split("_")
        name = parts[0]

        # parts
        # ['BloodsapBlade', '5001']
        # ['GoldCoinPurse']
        if len(parts) == 2:
            rarity_id = int(parts[1][0])
        elif len(parts) == 1:
            # some items have no rarity apparently
            rarity_id = 0

        position = data.get("slotId", -1)
        # Retrieve dimensions using ItemDataManager (was this the intention?)
        manager = ItemDataManager()
        _, width, height, _ = manager.get_item_image_path(name) or (None, 1, 1, None)

        return Item(name, rarity_id, position, width, height)

File: nodejs/item.js
// Node.js version of Python item.py
const { ItemDataManager } = require('./stashPreview');

class Item {
  constructor(name, rarity, position, width, height) {
    this.name = name;
    this.rarity = rarity;
    this.position = position;
    this.width = width;
    this.height = height;
  }

  static compare(a, b) {
    if (a.height !== b.height) return b.height - a.height;
    if (a.width !== b.width) return b.width - a.width;
    if (a.rarity !== b.rarity) return b.rarity - a.rarity;
    return a.name.localeCompare(b.name);
  }

  equals(other) {
    return (
      other &&
      this.name === other.name &&
      this.rarity === other.rarity &&
      this.position === other.position
    );
  }

  toString() {
    return `${this.rarity} ${this.name} ${this.position} ${this.width}X${this.height}`;
  }

  toDict() {
    return {
      name: this.name,
      rarity: this.rarity,
      position: this.position,
      width: this.width,
      height: this.height
    };
  }

  /**
   * Construct an Item from raw packet data using ItemDataManager for dimensions
   * @param {{ itemId: string, slotId?: number }} data
   * @returns {Promise<Item>}
   */
  static async fromDict(data) {
    const itemId = data.itemId || '';
    const designStr = 'DesignDataItem:Id_Item_';
    const parts = itemId.replace(designStr, '').split('_');
    const name = parts[0] || '';
    let rarity = 0;
    if (parts.length === 2) {
      rarity = parseInt(parts[1][0], 10);
    }
    const position = data.slotId ?? -1;

    const manager = new ItemDataManager();
    const imgInfo = await manager.getItemImagePath(name) || {};
    const width = imgInfo.width || 1;
    const height = imgInfo.height || 1;

    return new Item(name, rarity, position, width, height);
  }
}

module.exports = Item;